clang/lib/Sema/SemaDecl.cpp

```diff
@@ -11792,6 +11792,32 @@ static bool CheckMultiVersionFunction(Sema &S, FunctionDecl *NewFD,
                                          OldDecl, Previous);
 }
 
+static void CheckFunctionDeclarationAttributesUsage(Sema &S,
+                                                    FunctionDecl *NewFD) {
+  const bool is_pure = NewFD->hasAttr<PureAttr>();
+  const bool is_const = NewFD->hasAttr<ConstAttr>();
+
+  if (is_pure && is_const) {
+    S.Diag(NewFD->getLocation(), diag::warn_const_attr_with_pure_attr);
+    NewFD->dropAttr<PureAttr>();
+  }
+  if (is_pure || is_const) {
+    if (isa<CXXConstructorDecl>(NewFD)) {
```

Comment:
```suggestion
    if (isa<CXXConstructorDecl, DestructorDecl>(NewFD)) {
```
And destructors as well, yes? (Will need to update the diagnostic accordingly.)

Reply:
Destructors handled in 'function returning void' case and i think pure constructor is different from 'pure' function returning void while destructor is not

Reply:
I don't see why constructors and destructors should be handled differently. Neither has a way to specify a return value. My mental model for this is: pure and const only make sense on functions that are called for their return value. So anything which does not return a value is suspect to have either of these attributes. For ctor and dtor, there's no possible return value and so there is only one possible remedy (remove the attribute). For functions returning void, there are two possible remedies, remove the attribute or give the function a non-void return type.

Another situation worth thinking about is a function marked `noreturn` but I think those functions are typically declared with a `void` return type anyway and so I'm not certain we need special logic to diagnose this case.

Reply:
My point is that semantically a constructor has a return value, the user can think of a constructor as a function that returns a value. And in the future we may change the behavior so that marking the constructor 'pure' will make sense.

But with functions and destructors it will never make sense

Reply:
> My point is that semantically a constructor has a return value, the user can think of a constructor as a function that returns a value. And in the future we may change the behavior so that marking the constructor 'pure' will make sense.

I don't know if it's common for people to think of a constructor as a function that returns a value, but I can squint and see that line of thinking. CC @erichkeane for more opinions.

> This leads to just ignoring 'pure'/'const' on constructors without warning now to make possible change it in future (but gcc thinks it is UB)

I think we can change it in the future either way, this is about diagnostic wording.

Reply:
I don't agree that a constructor 'returns a value'.  Both a constructor and a destructor are the same in that they are functions that modify their 1st parameter (the this parameter) or 'throw'.  They are effectively the same thing.

I don't see how a constructor could ever make sense with 'pure', same as it doesn't make sense on destructors.

We should be handling them the same.

Reply:
> I don't agree that a constructor 'returns a value'. Both a constructor and a destructor are the same in that they are functions that modify their 1st parameter (the this parameter) or 'throw'. They are effectively the same thing.
> 
> I don't see how a constructor could ever make sense with 'pure', same as it doesn't make sense on destructors.
> 
> We should be handling them the same.

But user cannot separate the creation of an object and its construction from C++ code (not counting placement new)

```cpp
struct A {
... fields ...
  [[pure]] A(...args);
};

... 

A a = A(args...);
...
```
But on other side user can write:

```cpp
struct A { ... };

[[puer]] A A_constructor(...args);

```

And it will be correct, so, why i was hoping we let user create it less verbose in C++ interface

Reply:
>(not counting placement new)

Thats a big 'not counting' there...

`A a = A(args...);`

That isn't calling the constructor, that is effectively 'constructing an object, then copying it to the LHS (though we omit the copy)'.

`A a{args...};`  is also valid, which shows that it isn't the same.

Reply:
I wonder if start of a lifetime and end of the lifetime that is tied to constructor and destructor can be considered side effects in this context, preventing them from ever being pure functions.